### PR TITLE
Connecting receiver and 3 starts

### PR DESCRIPTION
**Pull request by sflemingo**
With this PR, an ubuntu machine can connect to the ublox receiver and start it with any of the three starts

### CODE REVIEW

File: examples/ublox_receiver_plugin/ublox_receiver_plugin.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  QWidget* createUI() override;
+  inline void initialize() override {}
+
+signals:
+  void configurationChanged();
+
+private:
+  SkydelNotifierInterface* m_skydelNotifier;
+  UbloxReceiverView* view;
+  Ublox* ubloxReceiver;
```
Comment (Rouiller) [inline examples/ublox_receiver_plugin/ublox_receiver_plugin.h]: Private class members should start with `m_`.

File: examples/ublox_receiver_plugin/ublox_receiver_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    // TODO: port is currently hardcoded... need to put code into CreateUbloxReceiver.cpp to find a port
+    // with a ublox receiver. Also need to change the ui (should ask for the baud rate but not the port)
+    CreateUbloxReceiverCommand command(baudRate, "/dev/ttyACM0");
+    ubloxReceiver = command.execute();
+    view->setReceiverStatus(ReceiverStatus::INACTIVE);
+  });
+
+  connect(view, &UbloxReceiverView::disconnectReceiver, [this]() {
+    m_skydelNotifier->notify("Disconnecting from the Ublox receiver");
+    this->ubloxReceiver->Disconnect();
```
Comment (Rouiller) [inline examples/ublox_receiver_plugin/ublox_receiver_plugin.cpp]: No need for `this->`, you can access directly `ubloxReceiver`(like you did with view on the next line).

File: examples/ublox_receiver_plugin/ublox_receiver_plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    view->setReceiverStatus(ReceiverStatus::NOT_CONNECTED);
+  });
+
+  connect(view, &UbloxReceiverView::startClicked, [this](ReceiverStartType startType) {
+    view->setReceiverStatus(ReceiverStatus::STARTING);
+    m_skydelNotifier->notify("Starting the Ublox receiver");
+    boost::thread startThread([this, startType]() {
+      ReceiverStartCommand command(this->ubloxReceiver);
+      command.execute(startType);
+      this->view->setReceiverStatus(ReceiverStatus::ACTIVE);
```
Comment (Rouiller) [inline examples/ublox_receiver_plugin/ublox_receiver_plugin.cpp]: Same here

File: examples/ublox_receiver_plugin/ublox_receiver_view.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+#include <QWidget>
+
+#include "receiver_enums.h"
+
+QT_BEGIN_NAMESPACE
+namespace Ui
+{
+class UbloxReceiverView;
+}
+QT_END_NAMESPACE
```
Comment (Rouiller) [inline examples/ublox_receiver_plugin/ublox_receiver_view.h]: I don't think you need QT_BEGIN_NAMESPACE and QT_END_NAMESPACE

File: examples/ublox_receiver_plugin/ublox_receiver_view.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+signals:
+  void connectReceiver(int baudRate);
+  void disconnectReceiver();
+  void startClicked(ReceiverStartType startType);
+
+private:
+  Ui::UbloxReceiverView* ui;
+  ReceiverStatus receiverStatus;
+  ReceiverStartType selectedStartType;
```
Comment (Rouiller) [inline examples/ublox_receiver_plugin/ublox_receiver_view.h]: Private class member should start with `m_`.

File: examples/ublox_receiver_plugin/ublox_receiver_view.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
+#include "ublox_receiver_view.h"
+
+#include "receiver_enums.h"
+#include "ui_ublox_receiver_view.h"
+
+UbloxReceiverView::UbloxReceiverView(QWidget* parent) : QWidget(parent), ui(new Ui::UbloxReceiverView)
+{
+  ui->setupUi(this);
+
+  selectedStartType = ReceiverStartType::NONE;
```
Comment (Rouiller) [inline examples/ublox_receiver_plugin/ublox_receiver_view.cpp]:
```suggestion
UbloxReceiverView::UbloxReceiverView(QWidget* parent) : QWidget(parent), ui(new Ui::UbloxReceiverView), selectedStartType(ReceiverStartType::NONE)
{
  ui->setupUi(this);
```

Comment (Rouiller) [reply examples/ublox_receiver_plugin/ublox_receiver_view.cpp]: Also `receiverStatus` is not explicitly initialised to a default value.

File: examples/ublox_receiver_plugin/ublox_receiver_view.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  setReceiverStatus(ReceiverStatus::NOT_CONNECTED);
+
+  connect(ui->startButton, &QPushButton::clicked, [this] {
+    ui->startButton->setEnabled(false);
+    emit startClicked(this->selectedStartType);
```
Comment (Rouiller) [inline examples/ublox_receiver_plugin/ublox_receiver_view.cpp]: No need for `this->`, applies for the rest of the file.

File: examples/ublox_receiver_plugin/ublox_receiver_view.cpp [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  connect(ui->startTypeSelect, qOverload<int>(&QComboBox::currentIndexChanged), [this](int index) {
+    this->selectedStartType = (ReceiverStartType)index;
+    if (index == 0)
+    {
+      ui->startButton->setEnabled(false);
+    }
+    else if (receiverStatus == ReceiverStatus::ACTIVE || receiverStatus == ReceiverStatus::INACTIVE)
+    {
+      ui->startButton->setEnabled(true);
+    }
```
Comment (Rouiller) [inline examples/ublox_receiver_plugin/ublox_receiver_view.cpp]: What happens when you are in the else case?

File: examples/ublox_receiver_plugin/ublox_receiver_view.cpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    if (this->receiverStatus == ReceiverStatus::NOT_CONNECTED)
+    {
+      int baudRate = ui->baudRateSelect->currentText().toInt();
+      emit connectReceiver(baudRate);
+    }
+    else
+    {
+      emit disconnectReceiver();
+    }
+  });
```
Comment (Rouiller) [inline examples/ublox_receiver_plugin/ublox_receiver_view.cpp]: For all the connects, when the lambda is more then a few lines, I think it's a good practice to create a function(slot) to make everything explicit.

File: examples/ublox_receiver_plugin/ublox_receiver_view.cpp [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+      break;
+    case ReceiverStatus::ACTIVE:
+      ui->connectReceiverButton->setText("Disconnect");
+      ui->baudRateSelect->setDisabled(true);
+      ui->receiverStatusLabel->setText("[Active]");
+      if (this->selectedStartType != ReceiverStartType::NONE)
+      {
+        ui->startButton->setDisabled(false);
+      }
+      break;
```
Comment (Rouiller) [inline examples/ublox_receiver_plugin/ublox_receiver_view.cpp]: Lots of code duplication code, maybe it's possible to make is more generic, I'll let you decide on that.

**Review by Rouiller — COMMENTED** [review]
Only a few comments and I only reviewed the plugin part, I wasn't sure what should be reviewed in the backend part. If you point me the files to review I can take a look at them later. Great job!

File: examples/ublox_receiver_plugin/ublox_receiver_view.cpp (hunk 8 above)
Comment (sflemingo) [reply examples/ublox_receiver_plugin/ublox_receiver_view.cpp]: The else case is when the receiverStatus is in a state where we cannot start the receiver (like NOT_CONNECTED), so we would not want to enable the startButton

File: examples/ublox_receiver_plugin/ublox_receiver_view.cpp (hunk 10 above)
Comment (sflemingo) [reply examples/ublox_receiver_plugin/ublox_receiver_view.cpp]: This gets fixed up a lot in the next PR (https://github.com/Team-Pinpoint/skydel-plug-ins/pull/10/files#diff-816a947cb51f09cea49067dd6538ea1a4a91347853c7ece6d697b190f2533a71), so I will keep this for now

**Review by Rouiller — APPROVED** [review]

**Pull request merged**